judge/judgedaemon.main.php

```diff
@@ -948,244 +974,238 @@ function judge(array $row)
-            'output_system' => rest_encode_file($testcasedir . '/system.out', $output_storage_limit),
-            'metadata' => rest_encode_file($testcasedir . '/program.meta', $output_storage_limit),
-            'output_diff'  => rest_encode_file($testcasedir . '/feedback/judgemessage.txt', $output_storage_limit)
-        );
-        $unsent_judging_runs[] = $new_judging_run;
-        $outstanding_data += strlen(var_export($new_judging_run, TRUE));
-
-        $now = now();
-        if (!$lastcase_correct
-            || ($now - $last_sent) >= $update_every_X_seconds
-            || $outstanding_data > $row['outputlimit'] * 1024) {
-           if (send_unsent_judging_runs($unsent_judging_runs, $row['judgingid']) === null) {
-               disable('problem', 'probid', $row['probid'], "uploading unsent judging runs failed", $row['judgingid'], (string)$row['cid']);
-               return;
-           }
-           $unsent_judging_runs = array();
-           $last_sent = $now;
-           $outstanding_data = 0;
-        }
-        logmsg(LOG_INFO, '  ' . ($result === 'correct' ? " \033[0;32m✔\033[0m" : " \033[1;31m✗\033[0m")
-            . ' ...done in ' . $runtime . 's, result: ' . $result);
-    } // end: for each testcase
-    if (!empty($unsent_judging_runs)) {
-        if (send_unsent_judging_runs($unsent_judging_runs, $row['judgingid']) === null) {
-            disable('problem', 'probid', $row['probid'], "uploading unsent judging runs failed", $row['judgingid'], (string)$row['cid']);
-            return;
-        }
+    if ($result === 'compare-error') {
+        logmsg(LOG_ERR, "comparing failed for compare script '" . $judgeTask['compare_script_id'] . "'");
+        $description = 'compare script ' . $judgeTask['compare_script_id'] . ' crashed';
+        disable('compare_script', 'compare_script_id', $judgeTask['compare_script_id'], $description, $judgeTask['judgetaskid']);
+        return false;
     }
 
-    // Sanity check: need to have had at least one testcase
-    if ($totalcases == 0) {
-        logmsg(LOG_WARNING, "No testcases judged for s$row[submitid]/j$row[judgingid]!");
+    $new_judging_run = array(
+        'runresult' => urlencode($result),
+        'runtime' => urlencode((string)$runtime),
+        'output_run'   => rest_encode_file($testcasedir . '/program.out', false),
+        'output_error' => rest_encode_file($testcasedir . '/program.err', $output_storage_limit),
+        'output_system' => rest_encode_file($testcasedir . '/system.out', $output_storage_limit),
+        'metadata' => rest_encode_file($testcasedir . '/program.meta', $output_storage_limit),
+        'output_diff'  => rest_encode_file($testcasedir . '/feedback/judgemessage.txt', $output_storage_limit),
+        'hostname' => $myhost,
+    );
+
+    if ($result === 'correct') {
+        // Post result back asynchronously. PHP is lacking multi-threading, so
```

Comment:
We do this async so that we can move on to the next testcase without waiting for the roundtrip http request to the server right?

Reply:
Yes, that's the idea. It saves quite some time to do this synchronously.